### PR TITLE
Add new backdown strategy to RetrySettings

### PR DESCRIPTION
**Pull request by borzun**
- Add new backdown strategy for RetrySettings (backdown_strategy filed in RetrySettings);
- Implement default implementation of `backdown_strategy`  - ExponentialBackdownStrategy;
- fix minor bug with retry attempt without any backdown policy.

Relates-to: OLPEDGE-1419

Signed-off-by: Bohdan Kurylovych <ext-bohdan.kurylovych@here.com>

### CODE REVIEW

File: olp-cpp-sdk-core/src/client/OlpClientSettings.cpp [hunk 1]
```diff
@@ -26,6 +27,18 @@ unsigned int DefaultBackdownPolicy(unsigned int milliseconds) {
   return milliseconds;
 }
 
+int FullJitterBackdownPolicy(int initial_backdown_period_msec,
+                             size_t retry_count) {
+  constexpr size_t kMaxRetries = 3;
```
Comment (andescu) [inline olp-cpp-sdk-core/src/client/OlpClientSettings.cpp]: Why max retries is only 3 and not what the user-specified?

Comment (borzun) [reply olp-cpp-sdk-core/src/client/OlpClientSettings.cpp]: Done - replaced with restriction on wait timeout in `OlpClient.cpp`

File: olp-cpp-sdk-core/include/olp/core/client/BackdownStrategy.h [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+   * @brief Function to compute next retry attemp wait time based on number of
+   * retries and initial backdown period.
+   *
+   * @param initial_backdown_period_msec Initial backdown period in
+   * milliseconds.
+   * @param retry_count Number of retries already made.
+   *
+   * @return Wait time in milliseconds for the next retry attempt.
+   */
+  int operator()(int initial_backdown_period_msec, size_t retry_count) {
```
Comment (r0busta) [inline olp-cpp-sdk-core/include/olp/core/client/BackdownStrategy.h]:
```suggestion
  std::chrono::milliseconds operator()(int initial_backdown_period_msec, size_t retry_count) {
```
I wouldn’t return a primitive type but a duration instead. It'll make API self-explanatory and hard use incorrectly.

Comment (r0busta) [inline olp-cpp-sdk-core/include/olp/core/client/BackdownStrategy.h]:
```suggestion
  int operator()(std::chrono::milliseconds initial_backdown_period_msec, size_t retry_count) {
```
I wouldn’t use a primitive type for the input but a duration instead. It'll make API self-explanatory and hard use incorrectly.

File: olp-cpp-sdk-core/include/olp/core/client/OlpClientSettings.h [hunk 3]
```diff
@@ -151,6 +152,16 @@ struct RetrySettings {
    */
   using BackdownPolicy = std::function<int(int)>;
 
+  /**
+   * @brief A strategy which takes the initial backdown period in milliseconds
+   * with current retry count, and returns what should be used for the next wait
+   * timeout.
```
Comment (r0busta) [inline olp-cpp-sdk-core/include/olp/core/client/OlpClientSettings.h]:
```{suggestion}
   * @brief A strategy for calculating retry timeouts based on
   * the initial backdown duration and retry count.
```
Would that read more straightforward and simpler?

File: olp-cpp-sdk-core/include/olp/core/client/OlpClientSettings.h [hunk 4]
```diff
@@ -178,11 +189,25 @@ struct RetrySettings {
...
    * @return The backdown policy.
    */
   BackdownPolicy backdown_policy = DefaultBackdownPolicy;
 
+  /**
+   * @brief The strategy used for retry attempts.
+   *
+   * The backdown strategy used to compute new wait time to retry failed
+   * requests. By default it is not specified and, if specified, it will used
+   * instead \ref backdown_policy.
```
Comment (r0busta) [inline olp-cpp-sdk-core/include/olp/core/client/OlpClientSettings.h]:
```{suggestion}
   * @brief Backdown strategy.
   *
   * Used for calculating retry timeouts for failed requests. By default, it is unset, and \ref backdown_policy is used.
```
Would that read simpler and more straightforward?

File: olp-cpp-sdk-core/include/olp/core/client/OlpClientSettings.h [hunk 5]
```diff
@@ -178,11 +189,25 @@ struct RetrySettings {
...
+   *
+   * @note You can use the \ref ExponentialBackdownStrategy from OLP SDK as new
```
Comment (r0busta) [inline olp-cpp-sdk-core/include/olp/core/client/OlpClientSettings.h]:
```suggestion
   * @note You can use the \ref ExponentialBackdownStrategy as the new
```
Redundant reference to the SDK

File: olp-cpp-sdk-core/src/client/OlpClient.cpp [hunk 6]
```diff
@@ -240,31 +240,44 @@ std::shared_ptr<http::NetworkRequest> OlpClient::CreateRequest(
...
-    int current_try, int current_backdown_period, const RetrySettings& settings,
+    int current_try, int current_backdown_period,
+    long long accumulated_wait_time, const RetrySettings& settings,
     const NetworkAsyncCallback& callback,
     const std::shared_ptr<http::NetworkRequest>& network_request,
     std::weak_ptr<http::Network> network,
     const std::weak_ptr<CancellationContext>& weak_cancel_context) {
   ++current_try;
   return [=](HttpResponse response) {
+    const long long max_wait_time =
```
Comment (r0busta) [inline olp-cpp-sdk-core/src/client/OlpClient.cpp]:
```suggestion
    const std::chrono::milliseconds max_wait_time =
```
Prefer abstract types over built-in types.

File: olp-cpp-sdk-core/src/client/OlpClient.cpp [hunk 7]
```diff
@@ -240,31 +240,44 @@ std::shared_ptr<http::NetworkRequest> OlpClient::CreateRequest(
...
+        settings.timeout * 1000ll;  // convert seconds to msecs
```
Comment (r0busta) [inline olp-cpp-sdk-core/src/client/OlpClient.cpp]:
```suggestion
        settings.timeout * 1000ll;
```
The comment is not needed if an abstract type is used.

File: olp-cpp-sdk-core/src/client/OlpClient.cpp [hunk 8]
```diff
@@ -240,31 +240,44 @@ std::shared_ptr<http::NetworkRequest> OlpClient::CreateRequest(
 }
 
 NetworkAsyncCallback GetRetryCallback(
-    int current_try, int current_backdown_period, const RetrySettings& settings,
+    int current_try, int current_backdown_period,
+    long long accumulated_wait_time, const RetrySettings& settings,
```
Comment (r0busta) [inline olp-cpp-sdk-core/src/client/OlpClient.cpp]:
```suggestion
    std::chrono::milliseconds accumulated_wait_time, const RetrySettings& settings,
```
Prefer abstract types over built-in types.

File: olp-cpp-sdk-core/src/client/OlpClient.cpp [hunk 9]
```diff
@@ -240,31 +240,44 @@ std::shared_ptr<http::NetworkRequest> OlpClient::CreateRequest(
...
       callback(std::move(response));
     } else {
       // TODO there must be a better way
-      std::this_thread::sleep_for(
-          std::chrono::milliseconds(current_backdown_period));
-
-      auto next_backdown_period =
-          settings.backdown_policy(current_backdown_period);
+      // Make sure that we don't wait longer than `timeout` in retry settings
+      const long long actual_wait_time = std::min<long long>(
```
Comment (r0busta) [inline olp-cpp-sdk-core/src/client/OlpClient.cpp]:
```suggestion
      const std::chrono::milliseconds actual_wait_time = std::chrono::milliseconds::min(
```
Prefer abstract types over built-in types.

File: olp-cpp-sdk-core/include/olp/core/client/OlpClientSettings.h [hunk 10]
```diff
@@ -151,6 +152,16 @@ struct RetrySettings {
    */
   using BackdownPolicy = std::function<int(int)>;
 
+  /**
+   * @brief A strategy which takes the initial backdown period in milliseconds
+   * with current retry count, and returns what should be used for the next wait
+   * timeout.
+   *
+   * Is superior to the \ref BackdownPolicy, as it computes wait time based on
+   * the retry count as well.
```
Comment (r0busta) [inline olp-cpp-sdk-core/include/olp/core/client/OlpClientSettings.h]: This note is not related to the type. I'd move it closer to the setting itself (i.e. `backdown_strategy`).

File: olp-cpp-sdk-core/src/client/OlpClient.cpp [hunk 11]
```diff
@@ -240,31 +240,44 @@ std::shared_ptr<http::NetworkRequest> OlpClient::CreateRequest(
...
+          current_backdown_period, max_wait_time - accumulated_wait_time);
+      std::this_thread::sleep_for(std::chrono::milliseconds(actual_wait_time));
+
+      int next_backdown_period = 0;
+      // New backdown policy should be superior to previous policy version:
```
Comment (r0busta) [inline olp-cpp-sdk-core/src/client/OlpClient.cpp]:
```suggestion
```
Redundant comment

File: olp-cpp-sdk-core/src/client/OlpClient.cpp [hunk 12]
```diff
@@ -240,31 +240,44 @@ std::shared_ptr<http::NetworkRequest> OlpClient::CreateRequest(
...
+      if (auto backdown_strategy = settings.backdown_strategy) {
+        next_backdown_period =
+            backdown_strategy(settings.initial_backdown_period, current_try);
+      } else if (auto backdown_policy = settings.backdown_policy) {
+        next_backdown_period = backdown_policy(current_backdown_period);
+      }
```
Comment (r0busta) [inline olp-cpp-sdk-core/src/client/OlpClient.cpp]: Move to a free function?

File: olp-cpp-sdk-core/src/client/OlpClient.cpp [hunk 13]
```diff
@@ -400,7 +414,12 @@ HttpResponse OlpClient::CallApi(
   auto response =
       SendRequest(network_request, settings_, retry_settings, context);
 
-  for (int i = 0; i < retry_settings.max_attempts && !context.IsCancelled();
+  // Make sure that we don't wait longer than `timeout` in retry settings
+  long long accumulated_wait_time = backdown_period;
+  const long long max_wait_time =
```
Comment (r0busta) [inline olp-cpp-sdk-core/src/client/OlpClient.cpp]: Prefer using abstract types instead of built-in ones.

File: olp-cpp-sdk-core/src/client/OlpClient.cpp [hunk 14]
```diff
@@ -400,7 +414,12 @@ HttpResponse OlpClient::CallApi(
...
+      retry_settings.timeout * 1000ll;  // convert seconds to msecs
```
Comment (r0busta) [inline olp-cpp-sdk-core/src/client/OlpClient.cpp]:
```suggestion
      retry_settings.timeout * 1000ll;
```
Can be removed if a self-explanatory abstract type used.

File: olp-cpp-sdk-core/src/client/OlpClient.cpp [hunk 15]
```diff
@@ -400,7 +414,12 @@ HttpResponse OlpClient::CallApi(
   auto response =
       SendRequest(network_request, settings_, retry_settings, context);
 
-  for (int i = 0; i < retry_settings.max_attempts && !context.IsCancelled();
+  // Make sure that we don't wait longer than `timeout` in retry settings
```
Comment (r0busta) [inline olp-cpp-sdk-core/src/client/OlpClient.cpp]:
```suggestion
```
Redundant comment.

File: olp-cpp-sdk-core/src/client/OlpClient.cpp [hunk 16]
```diff
@@ -411,14 +430,23 @@ HttpResponse OlpClient::CallApi(
...
+
     while (duration_to_sleep > 0 && !context.IsCancelled()) {
-      auto sleep_ms = std::min(1000, duration_to_sleep);
+      auto sleep_ms = std::min<long long>(1000, duration_to_sleep);
       std::this_thread::sleep_for(std::chrono::milliseconds(sleep_ms));
       duration_to_sleep -= sleep_ms;
     }
 
-    backdown_period = retry_settings.backdown_policy(backdown_period);
+    // New backdown policy should be superior to previous policy version:
```
Comment (r0busta) [inline olp-cpp-sdk-core/src/client/OlpClient.cpp]:
```suggestion
```
Redundant comment.

File: olp-cpp-sdk-core/src/client/OlpClient.cpp [hunk 17]
```diff
@@ -411,14 +430,23 @@ HttpResponse OlpClient::CallApi(
...
+    if (auto backdown_strategy = retry_settings.backdown_strategy) {
+      backdown_period =
+          backdown_strategy(retry_settings.initial_backdown_period, i);
+    } else if (auto backdown_policy = retry_settings.backdown_policy) {
+      backdown_period = backdown_policy(backdown_period);
+    }
```
Comment (r0busta) [inline olp-cpp-sdk-core/src/client/OlpClient.cpp]: Move to a free function?

File: olp-cpp-sdk-core/include/olp/core/client/BackdownStrategy.h [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+    const auto exponential_wait_time =
+        initial_backdown_period_msec * (1 << retry_count);
+
+    static thread_local std::mt19937 generator(std::random_device{}());
```
Comment (andescu) [inline olp-cpp-sdk-core/include/olp/core/client/BackdownStrategy.h]: static vars should be names similar to constants: `kGenerator`

File: olp-cpp-sdk-core/include/olp/core/client/BackdownStrategy.h [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+ * The actual formula can be described as:
+ * @code{.unparsed}
+ * wait_time = random_between(0, initial_backdown_period_msec * (2 ^
+ * retry_count))
+ * @endcode
+ */
+struct CORE_API ExponentialBackdownStrategy {
+ public:
+  /**
+   * @brief Function to compute next retry attemp wait time based on number of
```
Comment (andescu) [inline olp-cpp-sdk-core/include/olp/core/client/BackdownStrategy.h]:
```suggestion
   * @brief Function to compute next retry attempt wait time based on number of
```

File: olp-cpp-sdk-core/include/olp/core/client/OlpClientSettings.h [hunk 20]
```diff
@@ -178,11 +189,25 @@ struct RetrySettings {
 
   /**
    * @brief The backdown policy that should be used for the retry attempts.
+   * 
+   * @deprecated Use \ref backdown_strategy instead.
```
Comment (andescu) [inline olp-cpp-sdk-core/include/olp/core/client/OlpClientSettings.h]:
```suggestion
   * @deprecated Use \ref backdown_strategy instead. Will be removed by 06.2020.
```

File: olp-cpp-sdk-core/src/client/OlpClient.cpp (hunk 8 above)
Comment (andescu) [inline olp-cpp-sdk-core/src/client/OlpClient.cpp]: Please use int64_t or uint64_t instead of plain old data types.
Same for all other occurrences below.

File: olp-cpp-sdk-core/include/olp/core/client/BackdownStrategy.h [hunk 21]
```diff
@@ -0,0 +1,65 @@
...
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+#pragma once
+
+#include <olp/core/CoreApi.h>
+#include <random>
```
Comment (andescu) [inline olp-cpp-sdk-core/include/olp/core/client/BackdownStrategy.h]: Please separate system includes from project includes.
Same for olp-cpp-sdk-core/src/client/OlpClientSettings.cpp

File: olp-cpp-sdk-core/include/olp/core/client/BackdownStrategy.h [hunk 22]
```diff
@@ -0,0 +1,68 @@
...
+ * retry_count))
+ * @endcode
+ */
+struct CORE_API ExponentialBackdownStrategy {
+ public:
+  /**
+   * @brief Function to compute next retry attempt wait time based on number of
+   * retries and initial backdown period.
+   *
+   * @param initial_backdown_period_msec Initial backdown period in
```
Comment (r0busta) [inline olp-cpp-sdk-core/include/olp/core/client/BackdownStrategy.h]:
```suggestion
   * @param initial_backdown_period Initial backdown period.
```

File: olp-cpp-sdk-core/include/olp/core/client/BackdownStrategy.h [hunk 23]
```diff
@@ -0,0 +1,68 @@
...
+   * milliseconds.
```
Comment (r0busta) [inline olp-cpp-sdk-core/include/olp/core/client/BackdownStrategy.h]:
```suggestion
```

File: olp-cpp-sdk-core/src/client/OlpClient.cpp (hunk 15 above)
Comment (borzun) [reply olp-cpp-sdk-core/src/client/OlpClient.cpp]: I would argue about removal of this comment - I think it would be easier for reader to understand why we cap the max wait time.

File: olp-cpp-sdk-core/include/olp/core/client/BackdownStrategy.h [hunk 24]
```diff
@@ -0,0 +1,68 @@
...
+   * @param retry_count Number of retries already made.
+   *
+   * @return Wait time in milliseconds for the next retry attempt.
```
Comment (r0busta) [inline olp-cpp-sdk-core/include/olp/core/client/BackdownStrategy.h]:
```suggestion
   * @return Timeout for the next retry attempt.
```

File: olp-cpp-sdk-core/src/client/OlpClient.cpp (hunk 8 above)
Comment (borzun) [reply olp-cpp-sdk-core/src/client/OlpClient.cpp]: I used `std::chrono::milliseconds` instead.

File: olp-cpp-sdk-core/include/olp/core/client/BackdownStrategy.h [hunk 25]
```diff
@@ -0,0 +1,68 @@
...
+   */
+  std::chrono::milliseconds operator()(
+      std::chrono::milliseconds initial_backdown_period, size_t retry_count) {
+    const auto exponential_wait_time =
+        initial_backdown_period.count() * (1 << retry_count);
+
+    static thread_local std::mt19937 kGenerator(std::random_device{}());
```
Comment (r0busta) [inline olp-cpp-sdk-core/include/olp/core/client/BackdownStrategy.h]: Just asking. Would it make sense to set the seed?

File: olp-cpp-sdk-core/include/olp/core/client/OlpClientSettings.h [hunk 26]
```diff
@@ -151,6 +153,13 @@ struct RetrySettings {
    */
   using BackdownPolicy = std::function<int(int)>;
 
+  /**
+   * @brief A strategy for calculating retry timeouts based on
+   * the initial backdown duration in milliseconds and retry count.
```
Comment (r0busta) [inline olp-cpp-sdk-core/include/olp/core/client/OlpClientSettings.h]:
```suggestion
   * the initial backdown duration and retry count.
```

File: olp-cpp-sdk-core/src/client/OlpClient.cpp [hunk 27]
```diff
@@ -173,6 +173,21 @@ HttpResponse SendRequest(const http::NetworkRequest& request,
...
+  if (auto backdown_strategy = settings.backdown_strategy) {
+    return backdown_strategy(
+        std::chrono::milliseconds(settings.initial_backdown_period),
+        current_try);
+  } else if (auto backdown_policy = settings.backdown_policy) {
+    return std::chrono::milliseconds(
+        backdown_policy(current_backdown_period.count()));
+  } else {
+    return std::chrono::milliseconds::zero();
+  }
```
Comment (r0busta) [inline olp-cpp-sdk-core/src/client/OlpClient.cpp]: An unnecessary else statement. You can omit it.

File: olp-cpp-sdk-core/src/client/OlpClient.cpp [hunk 28]
```diff
@@ -240,31 +255,39 @@ std::shared_ptr<http::NetworkRequest> OlpClient::CreateRequest(
...
       callback(std::move(response));
     } else {
       // TODO there must be a better way
-      std::this_thread::sleep_for(
-          std::chrono::milliseconds(current_backdown_period));
+      const auto actual_wait_time = std::min(
+          current_backdown_period, max_wait_time - accumulated_wait_time);
+      std::this_thread::sleep_for(actual_wait_time);
+
+      const auto next_backdown_period =
```
Comment (r0busta) [inline olp-cpp-sdk-core/src/client/OlpClient.cpp]:
```suggestion
      const auto next_wait_time =
```
The variable name contradicts the function and other variable names

File: olp-cpp-sdk-core/tests/client/OlpClientTest.cpp [hunk 29]
```diff
@@ -368,6 +368,122 @@ TEST_P(OlpClientTest, RetryTimeExponential) {
   }
 }
 
+TEST_P(OlpClientTest, RetryWithExponentialBackdownStrategy) {
+  const std::chrono::milliseconds::rep kInitialBackdownPeriod = 100;  // msec
```
Comment (r0busta) [inline olp-cpp-sdk-core/tests/client/OlpClientTest.cpp]:
```suggestion
  const std::chrono::milliseconds::rep kInitialBackdownPeriod = 100; 
```

File: olp-cpp-sdk-core/tests/client/OlpClientTest.cpp [hunk 30]
```diff
@@ -368,6 +368,122 @@ TEST_P(OlpClientTest, RetryTimeExponential) {
...
+    const auto wait_time = olp::client::ExponentialBackdownStrategy()(
+        std::chrono::milliseconds(kInitialBackdownPeriod), retry_count);
+    wait_times.push_back(wait_time.count());
+    return wait_time;
+  };
+
+  // previous version of backdown policy shouldn't be called,
+  // if the new backdown policy was specified
+  client_settings_.retry_settings.backdown_policy = [](int period) -> int {
+    // we can't invoke FAIL directly from the callback:
```
Comment (r0busta) [inline olp-cpp-sdk-core/tests/client/OlpClientTest.cpp]: It's possible to add a non-fatal failure using ADD_FAILURE_AT which doesn't require a wrapping lambda.

File: olp-cpp-sdk-core/tests/client/OlpClientTest.cpp [hunk 31]
```diff
@@ -368,6 +368,122 @@ TEST_P(OlpClientTest, RetryTimeExponential) {
...
+
+  // verify that duration between retries matches actual wait time from
+  // backdown policy
+  for (size_t i = 1; i < timestamps.size(); ++i) {
+    ASSERT_GE((timestamps[i] - timestamps[i - 1]).count(), wait_times[i - 1]);
+  }
+}
+
+TEST_P(OlpClientTest, RetryTimeout) {
+  const auto kInitialBackdownPeriod = 400;  // msec
```
Comment (r0busta) [inline olp-cpp-sdk-core/tests/client/OlpClientTest.cpp]: Is it used anywhere else except for the line 443? I'd remove it if not

File: olp-cpp-sdk-core/tests/client/OlpClientTest.cpp [hunk 32]
```diff
@@ -368,6 +368,122 @@ TEST_P(OlpClientTest, RetryTimeExponential) {
...
+  const size_t kMaxRetries = 3;
+  const size_t kTimeout = 1;  // seconds
```
Comment (r0busta) [inline olp-cpp-sdk-core/tests/client/OlpClientTest.cpp]: Is it used anywhere else except for the line 445?  I'd remove it if not

File: olp-cpp-sdk-core/tests/client/OlpClientTest.cpp [hunk 33]
```diff
@@ -368,6 +368,122 @@ TEST_P(OlpClientTest, RetryTimeExponential) {
...
+                             olp::http::Network::Payload payload,
+                             olp::http::Network::Callback callback,
+                             olp::http::Network::HeaderCallback header_callback,
+                             olp::http::Network::DataCallback data_callback) {
+            current_attempt++;
+            // the test shouldn't reach the last retry due to timeout
+            // restrictions in retry settings
+            if (current_attempt == kSuccessfulAttempt) {
+              auto fail_helper = []() { FAIL(); };
+              fail_helper();
```
Comment (r0busta) [inline olp-cpp-sdk-core/tests/client/OlpClientTest.cpp]: Give ADD_FAILURE or ADD_FAILURE_AT a try. It will help to avoid this wrapping lambda.

File: olp-cpp-sdk-core/include/olp/core/client/OlpClientSettings.h (hunk 26 above)
Comment (borzun) [reply olp-cpp-sdk-core/include/olp/core/client/OlpClientSettings.h]: Disagree - I think it is good idea to emphasize that we work with milliseconds.

Comment (r0busta) [reply olp-cpp-sdk-core/include/olp/core/client/OlpClientSettings.h]: It's becoming bikeshedding :) So I'm okay with this. 
Just add that it's a bit against DRY (repeats what the type already says) and YAGNI (too many details) principles.

File: olp-cpp-sdk-core/include/olp/core/client/BackdownStrategy.h (hunk 25 above)
Comment (borzun) [reply olp-cpp-sdk-core/include/olp/core/client/BackdownStrategy.h]: Actually, I'm seeding the Mersenne Twister RNG with a seed value generated by `std::random_device`.  Note that random_device is a bit heavy-weight, as it can access OS resources (like reading https://en.wikipedia.org/wiki//dev/random file). 

Alternatively, we can use light-weighter approach, like:
-  seeding some magic seed value like 42;
- good'n'old `time(nullptr)`; 
- default seed value of Mersenne Twister (`std::mt19937::default_seed`);
- 'seed' a seed sequence (https://en.cppreference.com/w/cpp/numeric/random/seed_seq) for random generator, if we want more entropy, but it looks more like an over-engineering to me in such simple task as calculate next backdown strategy.

However, I have chosen the random device approach, as it is de-facto very popular among other examples that I saw on SO/C++ community regarding usage of C++ pRNG. 

P.S.: Also, after a bit of research on this topic, we can use a C++11-ish `srand()` approach - http://www.cplusplus.com/reference/random/minstd_rand/ instead Mersenne Twister approach, if we want a really simple pRNG. However, MT is more like default random numbers generator and should be used by default.

File: olp-cpp-sdk-core/include/olp/core/client/OlpClientSettings.h (hunk 26 above)
Comment (borzun) [reply olp-cpp-sdk-core/include/olp/core/client/OlpClientSettings.h]: Ok, I'll remove it. :)
